### PR TITLE
Migrate LoRA to ModelConfigTransform

### PR DESCRIPTION
**Pull request by jinsooihm**
Replace the LoRA `ModelConfigConverter` with `ModelConfigTransform`. 
- It must run after `ContextParallelTransform` because when the transform freezes a module, it makes a new subclass. 
This conflicts with `convert_config_type` inheritance requirement. 
- It conflicts with itself because of the parameter freezing. In other words, LoRA converter should not be run twice.
- Support extensibility to applying LoRA to other modules than those that subclass `Linear`. Specifically `GroupedLinear` in #4539. 

## Alternate design
An alternate design could be to separate freezing and LoRA as separate transforms. 
Then, we can make separate transforms for one that targets `Linear` and the other that targets `GroupedLinear`, and make them run after the freeze transform. But LoRA may be meaningless without freezing first, and the current `ModelConfigTransform` api does not have something like `requires` to enforce a transform to depend on the other.

## Test Plan

- `pytest tests/unit_tests/cpu/test_lora.py tests/unit_tests/cpu/test_transforms.py tests/unit_tests/cpu/test_torchft_trainer.py tests/unit_tests/cpu/test_integration_test_definitions.py -x`
  - 45 passed.

- Ran the `float8_emulate_lora` integration test for 10 training steps on 8 H100s with FSDP2, TP2, and PP2.
  - Passed successfully.

- Compared b21f7d43edc7 (`LoRAConverter`) against this change (`LoRATransform`) using `scripts/loss_compare.py` with `--debug.seed=42`, `--debug.deterministic` on `llama3_debugmodel_float8_emulate_lora_tp2_pp2`
  - Loss matched exactly for all 10 steps.
  - Gradient norm matched exactly for all 10 steps.

- Independently compared the unquantized debug-model configurations.
  - Config structure, all 81 state-dict tensors, parameter trainability, and forward output were bitwise identical.

### CODE REVIEW

File: torchtitan/config/transform/apply.py [hunk 1]
```diff
@@ -47,9 +47,9 @@ def _ordered(
 
 
 def _reject_conflicts(transforms: list[ModelConfigTransform]) -> None:
-    for transform in transforms:
-        for other in transforms:
-            if other is transform:
+    for i, transform in enumerate(transforms):
+        for j, other in enumerate(transforms):
+            if i == j:
```
Comment (fegin) [inline torchtitan/config/transform/apply.py:52]: Why this change?

File: torchtitan/config/transform/lora.py [hunk 2]
```diff
@@ -142,40 +55,165 @@ def _make_frozen_config(cfg: Module.Config) -> Module.Config:
...
+    The base transform supports ``Linear.Config``. Subclasses may extend
+    ``handlers`` for other projection types. Non-target modules are replaced
+    with dynamic frozen config subclasses that freeze direct parameters at
+    build time.
 
     When ``target_modules`` is None (default), every supported projection is
     converted. When specified, only configs whose FQN's last segment matches
     one of the entries are converted (e.g. ``["wq", "wv"]``).
+
+    This transform conflicts with itself because every application freezes all
```
Comment (fegin) [inline torchtitan/config/transform/lora.py:183]: It seems to be legit to restrict apply the same transform multiple times. Like if ContextParalllelTransform convert an inner attention to context parallel inner attention, the effect of second ContextParallelTransform is unknown. So it sounds like we may want to apply this restriction for all transforms?

File: torchtitan/config/transform/lora.py [hunk 3]
```diff
@@ -18,104 +19,16 @@
 from torchtitan.protocols.module import Module
 from torchtitan.protocols.sharding import ShardingConfig
 
-from .converter import ModelConfigConverter
+from .base import ModelConfigTransform
+from .context_parallel import ContextParallelTransform
```
Comment (fegin) [inline torchtitan/config/transform/lora.py:23]: So if the conflict graph has a cycle, we should get a circular import error.

File: torchtitan/config/transform/lora.py [hunk 4]
```diff
@@ -142,40 +55,165 @@ def _make_frozen_config(cfg: Module.Config) -> Module.Config:
...
-            raise ValueError(f"LoRA rank must be positive, got {config.rank}")
-        self.config = config
-        self.rank = config.rank
-        self.alpha = config.alpha
-        self.target_modules = (
-            set(config.target_modules) if config.target_modules is not None else None
-        )
+    run_after: ClassVar[tuple[type[ModelConfigTransform], ...]] = (
+        ContextParallelTransform,
+    )
```
Comment (tianyu-l) [inline torchtitan/config/transform/lora.py:190]: Please add comments why this requirement is there. In particular, I didn't get

> It must run after ContextParallelTransform because when the transform freezes a module, it makes a new subclass.

The target of CP is inner attention which doesn't have params anyway.

File: torchtitan/config/transform/lora.py [hunk 5]
```diff
@@ -142,40 +55,165 @@ def _make_frozen_config(cfg: Module.Config) -> Module.Config:
...
+    handlers: ClassVar[tuple[_LoRAHandler, ...]] = (_LinearLoRAHandler(),)
```
Comment (tianyu-l) [inline torchtitan/config/transform/lora.py:191]: related: for mxfp8 quantization to be applied on different modules, e.g. MoE and Linear, should we have multiple separate transform or multiple handlers. I don't think the "freezing twice" non-sense still exists, but are there other concerns?

File: torchtitan/config/transform/lora.py
Comment (tianyu-l) [inline torchtitan/config/transform/lora.py:1]: Maybe after the current work is done, please help migrate other converters as well. Thanks!

File: torchtitan/models/utils.py
Comment (fegin) [inline torchtitan/models/utils.py:30]: We may not be able to delete this function in this PR. `LMHeadCastConverter.Config` is also checked in the latest code. Please rebase your PR first.

File: torchtitan/config/transform/lora.py [hunk 6]
```diff
@@ -142,40 +55,165 @@ def _make_frozen_config(cfg: Module.Config) -> Module.Config:
...
+
+            def forward(self, input: torch.Tensor) -> torch.Tensor:
+                base_out = super().forward(input)
+                lora_out = self.lora_b(self.lora_a(input))
+                return base_out + self._lora_scaling * lora_out
+
+        LoRALinear.__name__ = f"LoRA{parent_cls.__name__}"
+        LoRALinear.__qualname__ = f"LoRA{parent_cls.__name__}"
+        self._class_cache[parent_cls] = LoRALinear
+        return LoRALinear
```
Comment (fegin) [inline torchtitan/config/transform/lora.py:152]: I think this is the main logic of LoRA, which should be move to `models/common/lora.py`.

You can create a base class, which make the main LoRA logic easy to understand.
```
class _LoRALinearMixin:
     def __init__(self, config) -> None:
           ...
 
     def forward(self, input: torch.Tensor) -> torch.Tensor:
           ...

     @staticmethod
     def _adapter_sharding(...)
```

And a new  util:
```
@functools.cache
def specialize_lora_linear(parent_cls: type[Linear]) -> type[Linear]:
   ...
```


Then the handler here just need to define `make_config`, which will call `specialize_lora_linear` and pass the parent class.  The main LoRA logic should not be hide inside the transform utils.

**Review by fegin — CHANGES_REQUESTED** [review]

File: torchtitan/config/transform/lora.py (hunk 4 above)
Comment (jinsooihm) [reply torchtitan/config/transform/lora.py:190]: `_make_frozen_config` makes a new class for every module. So if we apply LoRA transform, `FlexInnerAttention.Config` would become `FrozenFlexInnerAttention.Config`, which subclasses `FlexInnerAttention.Config`.

If we apply CP transform,  in `convert_config_type`, it will check for example `KVAllGatherCPFlexInnerAttention.Config` is a subclass of `FrozenFlexInnerAttention.Config` and raise.

I would guess it does not have to be this way.
1. I followed the `_make_frozen_config` behavior as it was originally implemented, but I don't see a reason why we should make these new Frozen classes. Is this perhaps for a possible implementation of checkpointing only the LoRA part? 
2. The `ContextParallelTransform` may be overly restrictive? What if I have a module with two different inner attentions, and I want to apply different CP kernels to them?

Comment (tianyu-l) [reply torchtitan/config/transform/lora.py:190]: Makes sense, thank you for explaining.

Re 1: https://github.com/pytorch/torchtitan/blob/main/torchtitan/config/transform/lora.py#L133-L134 I think this is not only for DCP, but a way to differentiate from original config class. Also it's subclass of original config class  https://github.com/pytorch/torchtitan/blob/main/torchtitan/config/transform/lora.py#L126. If CP transform uses `isinstance` https://github.com/pytorch/torchtitan/blob/main/torchtitan/config/transform/context_parallel.py#L42, it should work?

Re 2: Other than LoRA, which you are proposing using one transform + multiple handlers to cover both linear + experts with convincing reasons, we need stories for
- CP with multiple attention types, KDA / GDN / full @acisseJZhong @fegin 
- quantization: mxfp8 for linear + experts @anijain2305 @sanketpurandare

File: torchtitan/config/transform/apply.py (hunk 1 above)
Comment (jinsooihm) [reply torchtitan/config/transform/apply.py:52]: Related to this comment https://github.com/pytorch/torchtitan/pull/4684#discussion_r4012667861
this is because I made `LoRATransform.conflicts_with = (LoRATransform,)`, so the check must skip itself, but catch if there are two occurrences of `LoRATransform`.

File: torchtitan/config/transform/lora.py (hunk 2 above)
Comment (jinsooihm) [reply torchtitan/config/transform/lora.py:101]: I think we need to consider how we want to apply two different CP transforms. 
If we want to target layer 1,2,3 to use `KVAllGatherCPFlexInnerAttention` and layer 4 to use `UlyssesCPInnerAttention`, is it better to 
1. apply `ContextParallelTransform` twice? 
2. a single `ContextParallelTransform` apply them both by conditionally using one or the other?

Downside of 1 is that the logic of using `conflicts_with` becomes more complicated, because this would depend on the target modules. But at least for 1, even if we do not modify `conflicts_with` there will be a raise at `convert_config_type` so an erroneous application would be caught.

Comment (fegin) [reply torchtitan/config/transform/lora.py:101]: I prefer one transform and use some mapping (or the handler here?) to replace all attentions. Then we don't need to worry about `conflicts_with`.

File: torchtitan/config/transform/lora.py (hunk 4 above)
Comment (jinsooihm) [reply torchtitan/config/transform/lora.py:108]: Using `isinstance` could work, but I think it removes one safety check. `FrozenConfig.build` sets `param.requires_grad(False)`. If we replace without forcing inheritance, the behavior may be lost. It is fine for CP and LoRA, but perhaps there could be some more pathological examples, in which there is a silent drift? 

One may argue that the writer of a transform should take this into account and set `run_after` or `conflicts_with` correctly. But just like the situation of LoRA and CP, purely logically they can be applied in any order, but the implementation details prevent that, and `issubclass` catches that.

File: torchtitan/config/transform/lora.py (hunk 5 above)
Comment (jinsooihm) [reply torchtitan/config/transform/lora.py]: https://github.com/pytorch/torchtitan/pull/4684#discussion_r4018885130 
I think to make `conflicts_with` and `run_after` not too complicated and well defined, it would be better to keep one transform? But also this makes me wonder if it makes sense to apply mxfp8 qunatization on some modules, float8 on some other modules, and nvfp4 on the others.

File: torchtitan/config/transform/converter.py [hunk 7]
```diff
@@ -37,9 +37,8 @@ def convert(self, model_config: Module.Config) -> Module.Config:
 def validate_converter_order(
     converters: list[ModelConfigConverter.Config],
 ) -> None:
-    """Validate converter compatibility and ordering before model conversion."""
```
Comment (fegin) [inline torchtitan/config/transform/converter.py:40]: lol, the API name doesn't meet the description.

File: torchtitan/config/transform/lora.py (hunk 4 above)
Comment (fegin) [reply torchtitan/config/transform/lora.py:108]: Current CP transform is restricted. There is a TODO to support multiple attention cases. I have a design in my mind but it is not the same as the handler design in this PR. I don't know which one better. We can evaluate after we also let quantization be a transform and see if either design can support all and make a decision or propose a new design.

File: torchtitan/config/transform/lora.py [hunk 8]
```diff
@@ -142,40 +48,88 @@ def _make_frozen_config(cfg: Module.Config) -> Module.Config:
...
+
+    def __post_init__(self) -> None:
+        for index, handler in enumerate(self.handlers):
+            for earlier in self.handlers[:index]:
+                assert not issubclass(handler.config_type, earlier.config_type), (
+                    f"{type(handler).__qualname__} for "
+                    f"{handler.config_type.__qualname__} is shadowed by earlier "
+                    f"handler {type(earlier).__qualname__} for "
+                    f"{earlier.config_type.__qualname__}."
+                )
```
Comment (fegin) [inline torchtitan/config/transform/lora.py]: Let's use `raise ValueError`

File: torchtitan/config/transform/lora.py [hunk 9]
```diff
@@ -142,40 +48,88 @@ def _make_frozen_config(cfg: Module.Config) -> Module.Config:
...
-        self.config = config
-        self.rank = config.rank
-        self.alpha = config.alpha
-        self.target_modules = (
-            set(config.target_modules) if config.target_modules is not None else None
-        )
+    run_after: ClassVar[tuple[type[ModelConfigTransform], ...]] = (
+        ContextParallelTransform,
+    )
+    handlers: ClassVar[tuple[_LoRAHandler, ...]] = (_LinearLoRAHandler(),)
```
Comment (fegin) [inline torchtitan/config/transform/lora.py]: Okay, so your design is to let one Transform owns multiple handler and each handler matches the `config_type`. Then `handlers` shouldn't have the default value and shouldn't be a ClassVar. Instead, we should explicitly ask users to pass in the handlers. So on LoRATransform fits different modules. This is getting close to how are we going to support CP, except that CP also has Mask differences.

For this PR, remove `ClassVar`, expose `LinearLoRAHandler`, users have to explicitly pass the handlers.

File: torchtitan/models/common/lora.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+from torchtitan.models.common.decoder_sharding import dense_param_placement
+from torchtitan.models.common.linear import Linear
+from torchtitan.protocols.module import Module
+from torchtitan.protocols.sharding import ShardingConfig
+
+__all__ = ["specialize_lora_linear"]
+
+
+class _LoRALinearMixin:
+    """Add LoRA adapter parameters and computation to a linear module."""
```
Comment (fegin) [inline torchtitan/models/common/lora.py:31]: Do we still need "adapter'?

File: torchtitan/models/common/lora.py [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+            param_init={"weight": nn.init.zeros_},
+        ).build()
+
+    def forward(self, input_XI: torch.Tensor) -> torch.Tensor:
+        base_out_XO = super().forward(input_XI)  # type: ignore[misc]
+        lora_out_XO = self.lora_b(self.lora_a(input_XI))
+        return base_out_XO + self._lora_scaling * lora_out_XO
+
+    @staticmethod
+    def _adapter_sharding(
```
Comment (fegin) [inline torchtitan/models/common/lora.py:64]: I'm not sure if we want to remove `adapter`. I don't have a strong preference, but the `adapter` prefix seems not very correct now.

**Review by fegin — COMMENTED** [review]
Mostly good, please fix some nits and expose handlers and we can land.

File: torchtitan/config/transform/converter.py (hunk 7 above)
Comment (jinsooihm) [reply torchtitan/config/transform/converter.py:40]: Will change the name to `validate_converter_compatibility`

File: torchtitan/models/common/lora.py (hunk 10 above)
Comment (jinsooihm) [reply torchtitan/models/common/lora.py:31]: I guess A means adpatation so no.

Comment (jinsooihm) [reply torchtitan/models/common/lora.py:31]: But Codex seems to disagree and say 
> LoRA means Low-Rank Adaptation, not adoption. “LoRA adapter” is not strictly redundant: LoRA names the technique, while “adapter” names the added A/B parameter modules.
Recommendation: keep “LoRA adapter” where referring to those modules or parameters. It is standard and clearer than “LoRA parameters,” especially when distinguishing them from the frozen base model. No code change is needed.

Comment (jinsooihm) [reply torchtitan/models/common/lora.py:31]: Huggingface seems to use the term Tuner or Adapter 
https://huggingface.co/docs/peft/main/en/package_reference/tuners

File: torchtitan/config/transform/lora.py (hunk 5 above)
Comment (tianyu-l) [reply torchtitan/config/transform/lora.py]: makes sense, please leave comments in https://github.com/pytorch/torchtitan/pull/4541/changes#r4010691440

**Review by tianyu-l — APPROVED** [review]
sgtm, please wait for fegin approval

**Review by fegin — APPROVED** [review]
Thanks for the nice work
